components/renku_data_services/project/db.py

```diff
@@ -0,0 +1,123 @@
+"""Adapters for project database classes."""
+
+from __future__ import annotations
+
+from datetime import datetime, timezone
+from typing import Callable, NamedTuple, Tuple
+
+from sqlalchemy import func, select
+from sqlalchemy.ext.asyncio import AsyncSession
+
+import renku_data_services.base_models as base_models
+from renku_data_services import errors
+from renku_data_services.project import models
+from renku_data_services.project import orm as schemas
+
+
+class PaginationResponse(NamedTuple):
+    page: int
+    per_page: int
+    total: int
+    total_pages: int
+
+
+class ProjectRepository:
+    """Repository for project."""
+
+    def __init__(self, session_maker: Callable[..., AsyncSession]):
+        self.session_maker = session_maker  # type: ignore[call-overload]
+
+    async def get_projects(
+        self, user: base_models.APIUser, page: str, per_page: str
+    ) -> Tuple[list[models.Project], PaginationResponse]:
+        """Get all projects from the database."""
+        page = int(page)
+        per_page = int(per_page)
+
+        if page < 1:
+            raise errors.ValidationError(message=f"Parameter 'page' must be a natural number")
+        if per_page < 1 or per_page > 100:
+            raise errors.ValidationError(message=f"Parameter 'per_page' must be between 1 and 100")
+
+        async with self.session_maker() as session:
+            stmt = select(schemas.ProjectORM)
+            stmt = stmt.limit(per_page).offset((page - 1) * per_page)
+            stmt = stmt.order_by(schemas.ProjectORM.creation_date.desc())
+            result = await session.execute(stmt)
+            projects_orm = result.scalars().all()
+
+            stmt = select(func.count()).select_from(schemas.ProjectORM)
+            result = await session.execute(stmt)
+            n_total_elements = result.scalar()
+            total_pages, reminder = divmod(n_total_elements, per_page)
+            if reminder:
+                total_pages += 1
+
+            pagination = PaginationResponse(page, per_page, n_total_elements, total_pages)
```

Comment:
Yes, there are race conditions with pagination, but I do not think there is much we can do about that. I do not think this should be a big deal in practice.

Reply:
We can change the implementation to use a project id as the last or first item in the page - rather than only just using offsets. Then offsetting based on this constant IDs would prevent some of these issues. But I think we can let this be for now and come back to this later if we see it becoming a problem.